### ISSUE TITLE
IEngine: vectorized HRR binding operations (follow-up to #160) — ComplexPointwiseMultiply / PermutationGather / UnitPhaseCodebookInit / ComplexPhaseCoherenceDecode / HRRBindAccumulate

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#248]
## Summary

Follow-up to #160 (spectral/audio ops). The HRE Reversal Curse / Multi-Hop Reasoning / Associative Recall campaign (cycles 5–12 of the HRR research) has surfaced a new category of ops that we keep hand-rolling by stitching together 6+ `TensorPrimitivesCore` elementwise calls plus scalar permutation loops. These are small enough that per-call overhead dominates, and several (permutation gather, HRR unbind) have no vectorized path at all.

Filing the concrete ops here so future HRR cycles stop reinventing them.

---

## Measured bottlenecks (from HRE cycles 5–12, 2026-04-22 to -23)

| Bottleneck | Current cost | Root cause |
|---|---|---|
| HRR bind per training pair at D=16384 | 100 µs × 1M pairs × 4 scales = 400s | 6 chained elementwise ops + scalar perm gather |
| HRR unbind at inference at D=16384 | ~50 µs per query | Same 6-op stitching + conjugate flip |
| Scalar perm gather `out[d] = in[perm[d]]` | Serial ~10 µs at D=16384 | No SIMD path; breaks every hot loop it's in |
| Full-vocab phase-coherence decode at V=256, D=16384 | ~1 ms per query | `IEngine.TensorMatMul` works but requires interleaved [Re | Im] representation, adds copy overhead |
| Unit-phase codebook init at V=512, D=16384 | ~100 ms | Scalar `Math.Cos/Sin` loop per entry |

Empirical: cycle 12-v2 with D=16384 × 4 scales × 1M-byte corpus × 2 seeds took 30+ min and crashed the test host due to the per-bind overhead + allocation churn. After pre-computing permuted codebooks and pre-allocating scratch, cycle 12-v3 with D=8192 completed in ~80s per seed. Native IEngine ops would close another ~3–5x.

---

### 1. `ComplexPointwiseMultiply` (HIGHEST IMPACT — every HRR bind + unbind)

Currently hand-rolled as 6 `TensorPrimitives` elementwise calls:
```csharp
// c = a ⊙ b  (pointwise complex multiply)
TensorPrimitivesCore.Invoke...<MultiplyOperatorDouble>(aRe, bRe, t1);
TensorPrimitivesCore.Invoke...<MultiplyOperatorDouble>(aIm, bIm, t2);
TensorPrimitivesCore.Invoke...<SubtractOperatorDouble>(t1, t2, cRe);
TensorPrimitivesCore.Invoke...<MultiplyOperatorDouble>(aRe, bIm, t1);
TensorPrimitivesCore.Invoke...<MultiplyOperatorDouble>(aIm, bRe, t2);
TensorPrimitivesCore.Invoke...<AddOperatorDouble>(t1, t2, cIm);
```

Requested API:
```csharp
public void NativeComplexPointwiseMultiplyDouble(
    ReadOnlySpan<double> aRe, ReadOnlySpan<double> aIm,
    ReadOnlySpan<double> bRe, ReadOnlySpan<double> bIm,
    Span<double> cRe, Span<double> cIm,
    bool conjugateB = false);
```

`conjugateB=true` gives us `c = a ⊙ conj(b)` for HRR unbind. Setting this flag subtracts flip sign on the Im·Re pair — free in the kernel.

**Estimated win:** 1 kernel instead of 6 separate calls per bind. Eliminates temporary `t1/t2` spans. ~4x speedup on HRR memory construction. 30 min → ~8 min for cycle 12-scale builds.

### 2. `PermutationGather` — indexed gather for non-commutative binding

Currently hand-rolled:
```csharp
for (int d = 0; d < D; d++) outRe[d] = inRe[perm[d]];
```

Requested API:
```csharp
public void NativeGatherDouble(
    ReadOnlySpan<double> input,
    ReadOnlySpan<int> indices,
    Span<double> output);
```

This is a standard gather op (equivalent to numpy `a[perm]` or torch `torch.gather`), needed for Plate's original HRR trick of applying a fixed permutation to one operand to break commutativity. We pre-compute permuted codebooks at init to avoid this in the hot loop, but that doubles memory and we pay one gather per vocab entry at build.

**Estimated win:** 5–10x vs scalar loop; unlocks sparse-indexing optimizations in the tensors engine.

### 3. `UnitPhaseCodebookInit` — vectorized unit-phase code generation

Currently:
```csharp
for (int v = 0; v < V; v++)
    for (int d = 0; d < D; d++) {
        double phase = rng.NextDouble() * 2 * Math.PI;
        codeRe[v][d] = Math.Cos(phase);
        codeIm[v][d] = Math.Sin(phase);
    }
```

Requested API:
```csharp
public void NativeUnitPhaseCodebookDouble(
    Span<double> outRe, Span<double> outIm,  // flattened [V × D]
    int seed,
    int V, int D,
    bool kPsk = false, int k = 0);  // optional K-PSK quantization
```

**Estimated win:** ~10x via vectorized `SinCos` + `SimdRandom.NextDoubles` (which already exists per issue #157).

### 4. `ComplexPhaseCoherenceDecode` — fused matmul for V candidates

HRR decode computes `Re(Σ_d retrieved[d] · conj(code[v][d]))` for all V. This IS equivalent to a real matmul on the interleaved [Re | Im] representation:

Currently we wrap each query in a new `Tensor<double>([2D, 1])` and call `IEngine.TensorMatMul(codesT [V × 2D], query)` — works, but the per-call tensor construction adds overhead.

Requested API:
```csharp
public void NativeComplexPhaseCoherenceDecodeDouble(
    ReadOnlySpan<double> codesRe,  // flattened [V × D]
    ReadOnlySpan<double> codesIm,  // flattened [V × D]
    ReadOnlySpan<double> queryRe,  // [D]
    ReadOnlySpan<double> queryIm,  // [D]
    Span<double> scores);          // [V]
```

**Estimated win:** 2–3x vs current matmul + interleave layout trick. Removes the `new Tensor<double>([2D, 1])` allocation per query.

### 5. `ComplexMemoryAccumulateBind` — fused bind + accumulate

The most common HRR hot loop is:
```csharp
// memory += a_code ⊙ b_code_permuted
for all (a, b) in training: memory[d] += a[d] * b[perm[d]]  (complex)
```

This could be one kernel that handles the gather + multiply + accumulate + complex arithmetic:
```csharp
public void NativeHRRBindAccumulateDouble(
    ReadOnlySpan<double> keyCodebookRe, ReadOnlySpan<double> keyCodebookIm,  // [V × D]
    ReadOnlySpan<double> valPermCodebookRe, ReadOnlySpan<double> valPermCodebookIm,  // [V × D], pre-permuted
    ReadOnlySpan<int> keyIds, ReadOnlySpan<int> valIds,  // training pairs
    Span<double> memoryRe, Span<double> memoryIm);  // [D] accumulator
```

Processes N training bindings in one kernel call instead of N × 6 TensorPrimitives calls.

**Estimated win:** 5–10x on the HRR memory build — this is the single hottest loop in the HRE code.

---

## Motivation context

The HRE campaign has produced strong positive results with these ops hand-rolled (cycle 5: 100% associative recall zero params; cycle 6f: 10-hop recursive reasoning 100%; cycle 11d: Reversal Curse broken at 100% vs Transformer 0.3%). But cycle 12 (statistical LM at 1MB scale) is compute-bound on the hand-rolled inner loops — we haven't been able to run the full ablation sweep in a reasonable time.

Native IEngine ops would:
1. Enable cycle 12 scale sweeps (D=16384 × multi-scale × 1MB corpus) to finish in minutes, not hours
2. Make HRR primitives competitive with softmax attention on raw speed (we already beat it on params/cache/accuracy)
3. Remove the "hand-roll trap" where we keep slipping into scalar loops (see `feedback_hrr_vectorization_mandatory.md`)

Tagging this as HRE-track follow-up to #160; willing to implement any of these myself if the engine authoring guidelines allow.
